### PR TITLE
Add CodeQL Security Scan

### PR DESCRIPTION
**Pull request by KKelvinLo**
# Motivation
This PR is a follow-up to issue open-telemetry/oteps#144

[CodeQL](https://github.com/github/codeql-action) is GitHub's static analysis engine which scans repos for security vulnerabilities. As the project grows and we near GA it might be useful to have a workflow which checks for security vulnerabilities to ensure that every incremental change is following best development practices. Also passing basic security checks will also make sure that there aren't any glaring issues for our users.

## Changes
-  This PR adds [CodeQL](https://github.com/github/codeql-action) security checks to the repository
- After every run the workflow will upload the results to GitHub. 

#### Workflow Triggers
- daily cron job at 1:30 am
- workflow_dispatch (in the case that the maintainers want to trigger a manual security check)

* [X] `CHANGELOG.md` updated for non-trivial changes
* [] Unit tests have been added
* [] Changes in public API reviewed

### CODE REVIEW

**Comment by maxgolov** [comment]
@KKelvinLo - could you please take a look why the check hasn't been showing in the list of successful checks? You may want to adjust the template a bit for the results to show under the `Security tab`, maybe naming the step as `CodeQL`.

Template that's been missing the setup parts:
https://github.com/open-telemetry/opentelemetry-cpp/pull/654/files

**Comment by maxgolov** [comment]
Maybe it's all good already. Perhaps it'd show the results after merge?... 🤷

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    #        ┌───────────── minute (0 - 59)
+    #        │  ┌───────────── hour (0 - 23)
+    #        │  │ ┌───────────── day of the month (1 - 31)
+    #        │  │ │ ┌───────────── month (1 - 12 or JAN-DEC)
+    #        │  │ │ │ ┌───────────── day of the week (0 - 6 or SUN-SAT)
+    #        │  │ │ │ │
+    #        │  │ │ │ │
+    #        │  │ │ │ │
+    #        *  * * * *
+    - cron: '30 1 * * *'
```
Comment (maxgolov) [inline .github/workflows/codeql-analysis.yml:15]: No, we actually want to trigger this on `pull` and `push`.

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+name: "CodeQL Code Scan"
+
+on:
+  workflow_dispatch:
```
Comment (maxgolov) [inline .github/workflows/codeql-analysis.yml:4]:
```suggestion
  push:
    branches: [ main ]
  pull_request:
    # The branches below must be a subset of the branches above
    branches: [ main ]
```

**Review by lalitb — COMMENTED** [review]
Thank you for adding this. Will the CodeQL workflow fails if there are any security vulnerabilities, or it just uploads the result somewhere and we need to periodically keep checking it?

File: .github/workflows/codeql-analysis.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+name: "CodeQL Code Scan"
```
Comment (maxgolov) [inline .github/workflows/codeql-analysis.yml:1]:
```suggestion
name: "CodeQL"
```

That should show the code scan result under `Security` tab?

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (KKelvinLo) [reply .github/workflows/codeql-analysis.yml:15]: That's fair will change 👍

File: .github/workflows/codeql-analysis.yml (hunk 3 above)
Comment (KKelvinLo) [reply .github/workflows/codeql-analysis.yml:1]: Yep it if is enabled in the security tab, the action will post to the alerts

**Review by lalitb — APPROVED** [review]

File: .github/workflows/codeql-analysis.yml (hunk 2 above)
Comment (KKelvinLo) [reply .github/workflows/codeql-analysis.yml]: fixed

File: .github/workflows/codeql-analysis.yml (hunk 3 above)
Comment (maxgolov) [reply .github/workflows/codeql-analysis.yml]: Do we need to click on that in the `Security` tab or do we need to rename it to `CodeQL` ?

**Review by maxgolov — APPROVED** [review]

**Comment by KKelvinLo** [comment]
> Thank you for adding this. Will the CodeQL workflow fails if there are any security vulnerabilities, or it just uploads the result somewhere and we need to periodically keep checking it?

This workflow will post the results under the security tab, the workflow will pass

**Comment by KKelvinLo** [comment]
> @KKelvinLo - could you please take a look why the check hasn't been showing in the list of successful checks? You may want to adjust the template a bit for the results to show under the `Security tab`, maybe naming the step as `CodeQL`.
> 
> Template that's been missing the setup parts:
> https://github.com/open-telemetry/opentelemetry-cpp/pull/654/files

It should be good now - changed it from a cron schedule to on pull request and on push

**Review by ThomsonTan — APPROVED** [review]

**Comment by maxgolov** [comment]
> Code scanning results / CodeQL Timed out after 109m

It got crashed on recursive submodules?! 😀 

I guess we need to exclude `third_party` directory.

**Comment by alolita** [comment]
😀 @KKelvinLo can you please exclude the third-party directories. Let's make sure the workflow is robust. Thanks!

Thanks @maxgolov @ThomsonTan @lalitb for reviewing and merging!

**Comment by KKelvinLo** [comment]
> > Code scanning results / CodeQL Timed out after 109m
> 
> It got crashed on recursive submodules?! 😀
> 
> I guess we need to exclude `third_party` directory.

I added a bash command to remove the third_party directory after checking out the code, after digging through GitHub Actions documentation it doesn't look like there's an explicit exclude directory option.

**Comment by lalitb** [comment]
> > > Code scanning results / CodeQL Timed out after 109m
> > 
> > 
> > It got crashed on recursive submodules?! 😀
> > I guess we need to exclude `third_party` directory.
> 
> After digging through GitHub Actions documentation it doesn't look like there's an explicit exclude directory option, so I added a bash command to remove the third_party directory after checking out the code.

Thanks. Instead of removing the `third-party` folder later, we can also checkout the repo without submodules :).   For now, can you rebase the branch with main, as that will allow us to merge the PR

**Comment by KKelvinLo** [comment]
Rebased!

**Pull request merged**